### PR TITLE
[SPARK-14109][SQL] Fix HDFSMetadataLog to fallback from FileContext to FileSystem API

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

HDFSMetadataLog uses newer FileContext API to achieve atomic renaming. However, FileContext implementations may not exist for many scheme for which there may be FileSystem implementations. In those cases, rather than failing completely, we should fallback to the FileSystem based implementation, and log warning that there may be file consistency issues in case the log directory is concurrently modified.

In addition I have also added more tests to increase the code coverage.
## How was this patch tested?

Unit test. 
Tested on cluster with custom file system.

### CODE REVIEW

**Comment by tdas** [comment]
@marmbrus @zsxwing @JoshRosen

**Comment by tdas** [comment]
test this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 1]
```diff
@@ -196,4 +194,107 @@ class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)
...
+    }
+
+    override def deleteOnExit(path: Path): Unit = {
+      fc.deleteOnExit(path)
+    }
+  }
+
+  /** Implementation of FileManager using older FileSystem API */
+  class FileSystemManager(path: Path, hadoopConf: Configuration) extends FileManager {
+    private val fs = if (path.toUri.getScheme == null) {
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: You can just use `path.getFileSystem(hadoopConf)`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 2]
```diff
@@ -196,4 +194,107 @@ class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)
...
+
+    override def list(path: Path, filter: PathFilter): Array[FileStatus] = {
+      fs.listStatus(path, filter)
+    }
+
+    override def rename(srcPath: Path, destPath: Path): Unit = {
+      if (fs.exists(destPath)) {
+        throw new FileAlreadyExistsException(s"File already exists: $destPath")
+      }
+      fs.rename(srcPath, destPath)
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: Unlike `fc.rename`, `fs.rename` will return `false` if it fails. We should throw an exception for that.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLogSuite.scala [hunk 3]
```diff
@@ -17,16 +17,28 @@
...
 import org.scalatest.time.SpanSugar._
 
-import org.apache.spark.SparkFunSuite
+import org.apache.spark.{SparkConf, SparkFunSuite}
+import org.apache.spark.sql.execution.streaming.FakeFileSystem._
 import org.apache.spark.sql.test.SharedSQLContext
 
 class HDFSMetadataLogSuite extends SparkFunSuite with SharedSQLContext {
 
+
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLogSuite.scala]: nit: extra line

**Comment by zsxwing** [comment]
Looks pretty good except the `fs.rename` issue and some nits.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala (hunk 2 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: Yeah, that may be better.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 4]
```diff
@@ -196,4 +195,148 @@ class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)
...
+
+    /**
+     * Rename a path. Note that this implementation is not atomic.
+     * @throws FileNotFoundException if source path does not exist.
+     * @throws FileAlreadyExistsException if destination path already exists.
+     * @throws IOException if renaming fails for some unknown reason.
+     */
+    override def rename(srcPath: Path, destPath: Path): Unit = {
+      if (!fs.exists(srcPath)) {
+        throw new FileNotFoundException(s"Source path already exists: $srcPath")
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: nit: Source path already exists -> Cannot find ...

**Comment by tdas** [comment]
The last test failures were due to the changes in SharedSQLContext and TestSQLContext. The last commit should fix that.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 5]
```diff
@@ -196,4 +195,148 @@ class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)
...
+    def exists(path: Path): Boolean
+
+    /** Open a file for reading, or throw exception if it does not exist. */
+    def open(path: Path): FSDataInputStream
+
+    /** Create path, or throw exception if it already exists */
+    def create(path: Path): FSDataOutputStream
+
+    /**
+     * Atomically ename path, or throw exception if it cannot be done.
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: nit: ename -> rename

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 6]
```diff
@@ -196,4 +195,148 @@ class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)
...
+     * Should throw FileAlreadyExistsException if file already exists.
+     * Should throw FileNotFound exception if the file does not exist.
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: nit: `FileNotFound exception` -> FileNotFoundException

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 7]
```diff
@@ -196,4 +195,148 @@ class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)
...
+
+    /** Open a file for reading, or throw exception if it does not exist. */
+    def open(path: Path): FSDataInputStream
+
+    /** Create path, or throw exception if it already exists */
+    def create(path: Path): FSDataOutputStream
+
+    /**
+     * Atomically ename path, or throw exception if it cannot be done.
+     * Should throw FileAlreadyExistsException if file already exists.
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: nit `file` -> `srcPath`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala (hunk 6 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: nit: `file` -> `destPath`

**Comment by zsxwing** [comment]
LGTM except some nits

**Comment by tdas** [comment]
Alright I am merging this to master. Thanks @zsxwing

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 8]
```diff
@@ -196,4 +195,148 @@ class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)
...
   }
+
+  private def createFileManager(): FileManager = {
+    val hadoopConf = sqlContext.sparkContext.hadoopConfiguration
+    try {
+      new FileContextManager(metadataPath, hadoopConf)
+    } catch {
+      case e: UnsupportedFileSystemException =>
+        logWarning("Could not use FileContext API for managing metadata log file. The log may be" +
+          "inconsistent under failures.", e)
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: Can we remove this stack trace?  Its not helpful (the error is always thrown from `createFileSystem`.

**Pull request closed without merging**